drivers/hid/ipts/thread.h

```diff
@@ -0,0 +1,62 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Copyright (c) 2016 Intel Corporation
+ * Copyright (c) 2023 Dorian Stoll
+ *
+ * Linux driver for Intel Precise Touch & Stylus
+ */
+
+#ifndef IPTS_THREAD_H
+#define IPTS_THREAD_H
+
+#include <linux/completion.h>
+#include <linux/mutex.h>
+#include <linux/sched.h>
+
+/*
+ * This wrapper over kthread is neccessary, because calling kthread_stop makes it impossible
+ * to issue MEI commands from that thread while it shuts itself down. By using a custom
+ * boolean variable and a completion object, we can call kthread_stop only when the thread
+ * already finished all of its work and has returned.
+*/
+struct ipts_thread {
+	struct task_struct *thread;
+
+	bool should_stop;
+
+	struct mutex lock;
```

Comment:
You could replace those with set_bit() and test_bit() (https://www.kernel.org/doc/html/latest/core-api/kernel-api.html?highlight=set_bit#bit-operations) as full mutexes are a bit overkill for that task. Alternatives are atomics (which is probably better for explicit acquire/release semantics without needing rmb/wmb) or I guess you could even use `READ_ONCE()` and `WRITE_ONCE()` with memory barriers if you want to keep the bool.

So with bitops you'd do something like:
```c
int ipts_thread_stop(struct ipts_thread *thread)
{
	int ret = 0;

	if (!thread)
		return -EFAULT;

	if (!thread->thread)
		return 0;

	set_bit(IPTS_THREAD_STOP_BIT, &thread->flags);
	wmb();  // ensure bit is set before we start waiting

	wait_for_completion(&thread->done);
	ret = kthread_stop(thread->thread);

	thread->thread = NULL;
	thread->data = NULL;
	thread->threadfn = NULL;

	return ret;
}
```
and
```c
bool ipts_thread_should_stop(struct ipts_thread *thread)
{
	bool ret = false;

	if (!thread)
		return false;

	// optional: add rmb(); here to have "acquire" semantics, but that's not really necessary
	return test_bit(IPTS_THREAD_STOP_BIT, thread->flags);
}
```

Reply:
Would `smp_load_acquire` / `smp_store_release` work too? That could be the easiest solution, because it already seems to combine `READ_ONCE` and `WRITE_ONCE` with a barrier.

Reply:
Oh, I forgot about those. Yeah I think those should work too.

Reply:
Although... After checking, those have the barriers exactly the opposite way (which is what you more often want), but I think we only need the `wmb()` after setting the flag. If at all. I'm not entirely sure there how the semantics are for the `wait_for_completion()` call, but the point is that we want to make sure that the bit is set before waiting.

Reply:
Done